src/libcore/iter/mod.rs

```diff
@@ -1643,20 +1708,98 @@ impl<I: DoubleEndedIterator, U, F> DoubleEndedIterator for FlatMap<I, U, F> wher
 ///
 /// [`fuse()`]: trait.Iterator.html#method.fuse
 /// [`Iterator`]: trait.Iterator.html
-#[derive(Clone, Debug)]
 #[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
 #[stable(feature = "rust1", since = "1.0.0")]
-pub struct Fuse<I> {
-    iter: I,
-    done: bool
+pub struct Fuse<I: Iterator> {
+    iter: <I as SpecFuse>::T,
+}
+
+impl<I: Iterator> Fuse<I> {
+    fn new(iter: I) -> Self {
+        Fuse { iter: <I as SpecFuse>::T::new(iter) }
+    }
 }
 
+#[unstable(feature = "fused", issue = "35602")]
+impl<I> FusedIterator for Fuse<I> where I: Iterator {}
+
 #[stable(feature = "rust1", since = "1.0.0")]
-impl<I> Iterator for Fuse<I> where I: Iterator {
-    type Item = <I as Iterator>::Item;
+impl<I: Iterator + Clone> Clone for Fuse<I> {
+    fn clone(&self) -> Self {
+        Fuse { iter: self.iter.spec_clone() }
+    }
+}
 
-    #[inline]
-    fn next(&mut self) -> Option<<I as Iterator>::Item> {
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<I: Iterator + fmt::Debug> fmt::Debug for Fuse<I> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.iter.spec_debug(f)
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<I: Iterator> Iterator for Fuse<I> {
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<I::Item> {
+        self.iter.next()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<I: DoubleEndedIterator> DoubleEndedIterator for Fuse<I> {
+    fn next_back(&mut self) -> Option<I::Item> {
+        self.iter.spec_next_back()
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<I: ExactSizeIterator> ExactSizeIterator for Fuse<I> {
+    fn len(&self) -> usize {
+        self.iter.spec_len()
+    }
+}
+
+trait SpecFuse: Sized + Iterator {
+    type T: SpecFuseImpl<Iter = Self> + Iterator<Item=Self::Item>;
+}
+
+unsafe trait SpecFuseImpl: Sized {
```

Comment:
Because I was going to use `spec_get_unchecked` in a `TrustedRandomAccess` implementation but forgot to actually add the `TrustedRandomAccess` implementation...